packages/@aws-cdk/aws-stepfunctions/README.md

```diff
@@ -1,2 +1,322 @@
+choice.afterwards().next(shipTheItem);
+```
+
+If your `Choice` doesn't have an `otherwise()` and none of the conditions match
+the JSON state, a `NoChoiceMatched` error will be thrown. Wrap the state machine
+in a `Parallel` state if you want to catch and recover from this.
+
+### Parallel
+
+A `Parallel` state executes one or more subworkflows in parallel. It can also
+be used to catch and recover from errors in subworkflows.
+
+```ts
+const parallel = new stepfunctions.Parallel(this, 'Do the work in parallel');
+
+// Add branches to be executed in parallel
+parallel.branch(shipItem);
+parallel.branch(sendInvoice);
+parallel.branch(restock);
+
+// Retry the whole workflow if something goes wrong
+parallel.retry({ maxAttempts: 1 });
+
+// How to recover from errors
+parallel.onError(sendFailureNotification);
+
+// What to do in case everything succeeded
+parallel.next(closeOrder);
+```
+
+### Succeed
+
+Reaching a `Succeed` state terminates the state machine execution with a
+succesful status.
+
+```ts
+const success = new stepfunctions.Succeed(this, 'We did it!');
+```
+
+### Fail
+
+Reaching a `Fail` state terminates the state machine execution with a
+failure status. The fail state should report the reason for the failure.
+Failures can be caught by encompassing `Parallel` states.
+
+```ts
+const success = new stepfunctions.Fail(this, 'Fail', {
+    error: 'WorkflowFailure',
+    cause: "Something went wrong"
+});
+```
+
+## Task Chaining
+
+To make defining work flows as convenient (and readable in a top-to-bottom way)
+as writing regular programs, it is possible to chain most methods invocations.
+In particular, the `.next()` method can be repeated. The result of a series of
+`.next()` calls is called a **Chain**, and can be used when defining the jump
+targets of `Choice.on` or `Parallel.branch`:
+
+```ts
+const definition = step1
+    .next(step2)
+    .next(choice
+        .on(condition1, step3.next(step4).next(step5))
+        .otherwise(step6)
+        .afterwards())
+    .next(parallel
+        .branch(step7.next(step8))
+        .branch(step9.next(step10)))
+    .next(finish);
+
+new stepfunctions.StateMachine(this, 'StateMachine', {
+    definition,
+});
+```
+
+If you don't like the visual look of starting a chain directly off the first
+step, you can use `Chain.start`:
+
+```ts
+const definition = stepfunctions.Chain
+    .start(step1)
+    .next(step2)
+    .next(step3)
+    // ...
+```
+
+## State Machine Fragments
+
+It is possible to define reusable (or abstracted) mini-state machines by
+defining a construct that implements `IChainable`, which requires you to define
+two fields:
+
+* `startState: State`, representing the entry point into this state machine.
+* `endStates: INextable[]`, representing the (one or more) states that outgoing
+  transitions will be added to if you chain onto the fragment.
+
+Since states will be named after their construct IDs, you may need to prefix the
```

Comment:
No; the State ID also serves as a human-readable "caption" (it's the only user-visible description of a state) and people are going to look at state machine diagrams, PLUS it will have to be interpreted by the eventual tooling that's going to look at `DescribeExecutionHistory`.

So awkward as it is, I do want to offer full control of state IDs out of the box.

Reply:
If you use the construct path, it will only consist of human readable IDs (`/my/beautiful/task`)

Reply:
Can we allow people to customize but also offer a reasonable default that won't trip them off?

Reply:
I get the default, but I'm not sure how to build the API would look like that gives you the ability to override those defaults with enough flexibility; given that the constructs in between could be *anything* we can't assume they have awareness of States so we can't configure on the inbetween constructs.

I actually used to have that (`StateMachineFragments` in the upstream path would add to the name, others wouldn't) but (a) the automatic naming wasn't particularly well-received by SFN and (b) it ties construct parents to behavior, which is something we explicitly didn't want to do.

I'll pose the question again if just defaulting IDs is good enough. I agree it does simplify things.

Reply:
Also, the default behavior by going from a common ancestor between the state machine and the state also has some sharp edges. Do we calculate from the shared ancestor between ANY state in the tree and the state machine? If so, adding a state from a different part of the hierarchy has the potential to change the name of ALL STATES in the state machine.

If we do it on a per-state-object basis, names might still conflict if you add states from parts of the hierarchy that are higher up the tree than the state machine is (since the shared parent will effectively always be the parent of the state).

Reply:
But what do you suppose a rename API would look like?

It can't be on the state object, because the state might be instantiated multiple times (via a reusable parent).

It can't be on the parent because that might just be a `cdk.Construct` which is unaware of any step functions internals.

It COULD be on the StateMachine, but that's the same mechanism as stack-renames, and feels very after-the fact and not composable. Also, I will guarantee you that *everyone* will want to rename *all states* if we pick `construct.path` as a default. Then it's no longer an optional tweak, it becomes part of the recommended API.

Reply:
I think I understand your concerns. So how about we just use `construct.path` (or even something like `${construct.uniqueId}.` as the default `prefixState` for `StateMachineFragment`s. So by default, the usage of fragments is safe and doesn't require users to "remember" to call `prefixState`. If they want to customize, they can call `prefixState` with a custom prefix.